src/Compilers/CSharp/Portable/Binder/Binder_Conversions.cs

```diff
@@ -1093,6 +1094,14 @@ static ErrorCode getRefMismatchErrorCode(TypeKind type)
             MethodSymbol selectedMethod = conversion.Method;
 
             var location = syntax.Location;
+
+            if (!isAddressOf)
```

Comment:
>if (!isAddressOf) [](start = 12, length = 17)

Moving this entire block to the start of this method seems unnecessary. It might result in additional diagnostics for scenarios where we already report errors. But unless this catches error cases that previously compiled without errors, we should avoid changing the set of diagnostics reported.

Consider moving immediately above the `if (selectedMethod.HasUnsafeParameter ...) { ... }` block instead.

Reply:
I believe we did have a bug here with `UnmanagedCallersOnly` verification, but I think we should be consistent about reporting cascaded attributes. Either we should always report `UnmanagedCallersOnly` and `Obsolete` with the other errors, or we should never do so. Consider instead a structure like this:
```cs
if ((selectedMethod.HasUnsafeParameter() || selectedMethod.ReturnType.IsUnsafe()) && ReportUnsafeIfNotAllowed(syntax, diagnostics))
{
    return true;
}

if (!isAddressOf)
{
    ReportDiagnosticsIfUnmanagedCallersOnly(diagnostics, selectedMethod, location, isDelegateConversion: true);
}
ReportDiagnosticsIfObsolete(diagnostics, selectedMethod, syntax, hasBaseReceiver: false);
```

Reply:
@333fred With this structure, the `Obsolete` and `UnmanagedCallersOnly` warnings won't be reported if unsafe errors are reported. I think it's better if the warning existed with the error.

Reply:
> With this structure, the Obsolete and UnmanagedCallersOnly warnings won't be reported if unsafe errors are reported. I think it's better if the warning existed with the error.

This is correct. IMO, we should be consistent and either always report Obsolete/UnmanagedCallersOnly (ie, moving this to the front of the method), or only every report these errors when there are no others (ie, moving them to the bottom). Currently, we're just inconsistent, sometimes reporting these errors when other errors are present, and sometimes not.

Reply:
@333fred I currently have both of them together with before the unsafe checks

Reply:
> @333fred I currently have both of them together with before the unsafe checks

I understand. And that behavior is the inconsistent behavior I'm talking about. We should either always report these, no matter what other errors might be reported, or only report these when there are no other errors.

Reply:
@333fred Got it. I thought they were always reported.

Should I move the warnings check above just above the following IsConditional check, so they'are always reported?

```
           if (selectedMethod.IsConditional)
            {
                // CS1618: Cannot create delegate with '{0}' because it has a Conditional attribute
                Error(diagnostics, ErrorCode.ERR_DelegateOnConditional, location, selectedMethod);
                return true;
            }

Reply:
@cston preferred to have this not be reported if other errors are reported, so I'll defer to his preference. Let's move this to the end of the method, only reported if no other errors are reported.

Reply:
@333fred The checks were originally at the end of the method. I moved them based on @cston's comment (https://github.com/dotnet/roslyn/pull/47332#discussion_r492318014)

Reply:
I'm aware. That comment was about a bug in the unsafe reporting logic. With the suggested code I gave [here](https://github.com/dotnet/roslyn/pull/47332#discussion_r492366626), that is no longer an issue.